### PR TITLE
[next] Wrap all generated CSS with a cascade layer (WIP)

### PR DESCRIPTION
**Pull request by Janpot**
One way to fix https://github.com/mui/pigment-css/issues/199

* Updated [stackblitz](https://stackblitz.com/edit/github-xv6opg-fggtsn?file=package.json) for https://github.com/mui/pigment-css/issues/199
* Updated [codesandbox](https://codesandbox.io/p/github/wh5938316/mui-v6-pigment-css/csb-ymg78v/draft/floral-dream) for https://github.com/mui/material-ui/issues/43529.

Notes:

* Does this have unwanted side-effects?
* TODO: sourcemaps. Can we get away with them shifting down 1 line for the sake of a quick fix? => let's run with it for now.
* Does this belong in `wyw-in-js`?

Closes https://github.com/mui/material-ui/pull/43705

### CODE REVIEW

**Comment by DiegoAndai** [comment]
> Does this have unwanted side-effects?

We have visual regression tests running with Pigment on the Material UI repo, we can create a draft PR pointing to these builds to check if anything unexpected is happening there. It probably won't cover all edge cases but it would be helpful.

> Can we get away with them shifting down 1 line for the sake of a quick fix?

**Comment by DiegoAndai** [comment]
> Does this have unwanted side-effects?

We have visual regression tests running with Pigment on the Material UI repo, we can create a draft PR pointing to these builds to check if anything unexpected is happening there. It probably won't cover all edge cases but it would be helpful.

> Can we get away with them shifting down 1 line for the sake of a quick fix?

Why do you consider this to be a quick fix?

If it is a quick fix, can users implement it on their side? We could propose it as a workaround until we agree on a definitive solution.

> Does this belong in wyw-in-js?

If it is, let's try to fix it on our side anyway. We can open an issue on their repo and refactor it later, but let's not block a fix on our side.

**Comment by Janpot** [comment]
> Why do you consider this to be a quick fix?

Because it's just rewriting the css without considering the source maps. This means that the links in devtools will be off by one line. It may be quite a bit more involved to write this in a way that sourcemaps are transformed as well. It all comes down to how much effort we want to pour down into fixing this.

> If it is a quick fix, can users implement it on their side?

I don't see how.

**Comment by brijeshb42** [comment]
I am not sure how fool proof the fix is and if wouldn't cascade into some other bug.
Let me check this change with some calls to globalCss api.

**Comment by Janpot** [comment]
What do you mean with "globalCss api"? 

I've tested by adding a `./global.css` overriding some properties on [Mui classes](https://mui.com/material-ui/api/button/#classes). That still works since unlayered styles are more specific than layered styles.

File: packages/pigment-css-unplugin/src/index.ts [hunk 1]
```diff
@@ -284,6 +283,19 @@ export const plugin = createUnplugin<PigmentOptions, true>((options) => {
         }
 
         let { cssText } = result;
+
+        const slug = slugify(cssText);
```
Comment (brijeshb42) [inline packages/pigment-css-unplugin/src/index.ts:287]: Can you move this at the end of the `if (isNext)` check at line https://github.com/mui/pigment-css/blob/master/packages/pigment-css-unplugin/src/index.ts#L301

since we mainly want to do this for next.js and not webpack.

**Comment by brijeshb42** [comment]
> Does this have unwanted side-effects?
We'll have to see. It fixes the current bug.

> TODO: sourcemaps. Can we get away with them shifting down 1 line for the sake of a quick fix? => let's run with it for now.

Sourcemaps are working without any issues since the linkage is through the generated class names which we are not touching.

> Does this belong in wyw-in-js?

Not really since wyw does not have any nextjs specific solution. If it had `@wyw/nextjs`, then yes.

**Review by brijeshb42 — APPROVED** [review]
Thanks for the fix.
